### PR TITLE
[test] Make the canary image pull logic scalable

### PR DESCRIPTION
**Pull request by shantanutrip**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
The canaries did not scale properly for AG images. The logic of fetching the AG images was not able to handle the python version correctly. This PR helps handle that. Additionally, it makes the logic scalable for every other framework, so that we do not have to make changes every time.

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/test_utils/__init__.py [hunk 1]
```diff
@@ -862,6 +863,9 @@ def parse_canary_images(framework, region):
                 versions_counter[version]["tr"] = True
             elif "inf" in tag_str:
                 versions_counter[version]["inf"] = True
+            
+            if match.group(3):
```
Comment (arjkesh) [inline test/test_utils/__init__.py:867]: you don't need this condition, right - it's not an optional parameter in the regex

Comment (shantanutrip) [reply test/test_utils/__init__.py]: Done, thanks

File: test/test_utils/__init__.py [hunk 2]
```diff
@@ -862,6 +865,8 @@ def parse_canary_images(framework, region):
                 versions_counter[version]["tr"] = True
             elif "inf" in tag_str:
                 versions_counter[version]["inf"] = True
+            
+            pre_populated_py3_version[version] = match.group(3)
```
Comment (arjkesh) [inline test/test_utils/__init__.py:869]: so sometimes we might do a framework release with multiple python versions. If there are multiple python versions associated with a fw version, we may want to run canaries on all python versions. thus we can probably make this a list and iterate through while appending to the images list

File: test/test_utils/__init__.py [hunk 3]
```diff
@@ -836,21 +836,24 @@ def parse_canary_images(framework, region):
...
+        "pytorch": rf"pt(-sagemaker)?{customer_type_tag}-(\d+.\d+)\S*(py\d+)",
+        "huggingface_pytorch": r"hf-\S*pt(-sagemaker)?-(\d+.\d+)\S*(py\d+)",
+        "huggingface_tensorflow": r"hf-\S*tf(-sagemaker)?-(\d+.\d+)\S*(py\d+)",
+        "autogluon": r"ag(-sagemaker)?-(\d+.\d+)\S*(py\d+)",
     }
 
     repo = git.Repo(os.getcwd(), search_parent_directories=True)
 
     versions_counter = {}
+    pre_populated_py3_version = {}
```
Comment (arjkesh) [inline test/test_utils/__init__.py:850]: this isn't always explicitly py3 right - the regex is general for python version

File: test/test_utils/__init__.py [hunk 4]
```diff
@@ -836,21 +836,24 @@ def parse_canary_images(framework, region):
...
-        "pytorch": rf"pt(-sagemaker)?{customer_type_tag}-(\d+.\d+)",
-        "huggingface_pytorch": r"hf-\S*pt(-sagemaker)?-(\d+.\d+)",
-        "huggingface_tensorflow": r"hf-\S*tf(-sagemaker)?-(\d+.\d+)",
-        "autogluon": r"ag(-sagemaker)?-(\d+.\d+)",
+        "tensorflow": rf"tf(-sagemaker)?{customer_type_tag}-(\d+.\d+)\S*(py\d+)",
+        "mxnet": rf"mx(-sagemaker)?{customer_type_tag}-(\d+.\d+)\S*(py\d+)",
+        "pytorch": rf"pt(-sagemaker)?{customer_type_tag}-(\d+.\d+)\S*(py\d+)",
+        "huggingface_pytorch": r"hf-\S*pt(-sagemaker)?-(\d+.\d+)\S*(py\d+)",
+        "huggingface_tensorflow": r"hf-\S*tf(-sagemaker)?-(\d+.\d+)\S*(py\d+)",
+        "autogluon": r"ag(-sagemaker)?-(\d+.\d+)\S*(py\d+)",
```
Comment (arjkesh) [inline test/test_utils/__init__.py:844]: you may want to add the "-" in case we have a package in the future that interferes with this

File: test/test_utils/__init__.py [hunk 5]
```diff
@@ -876,50 +883,57 @@ def parse_canary_images(framework, region):
     framework_versions = versions if len(versions) < 4 else versions[:3]
     dlc_images = []
     for fw_version in framework_versions:
-        py3_version = get_canary_default_tag_py3_version(framework, fw_version)
```
Comment (arjkesh) [inline test/test_utils/__init__.py:879]: this is probably the only usage of this function, so you can prob remove the function too

File: test/test_utils/__init__.py [hunk 6]
```diff
@@ -876,50 +883,57 @@ def parse_canary_images(framework, region):
...
-        if customer_type == "e3":
-            dlc_images += [f"{img}-e3" for img in images[framework]]
+        if fw_version in pre_populated_py3_version:
+            py3_versions = pre_populated_py3_version[fw_version]
         else:
-            dlc_images += images[framework]
+            ## Keeping it here for legacy releases. Based on the new implementation
+            ## the tags always have py3 version appended to it and hence the 
+            ## get_canary_default_tag_py3_version might not be required. 
+            py3_versions = [get_canary_default_tag_py3_version(framework, fw_version)]
```
Comment (arjkesh) [inline test/test_utils/__init__.py:892]: it seems like there is no case where this will be hit - let's remove rather than keeping it around

File: test/test_utils/__init__.py [hunk 7]
```diff
@@ -876,50 +883,57 @@ def parse_canary_images(framework, region):
...
+        
+        for py3_version in py3_versions:
```
Comment (arjkesh) [inline test/test_utils/__init__.py:894]: once again, not necessarily py3 version right - just py version

File: test/test_utils/__init__.py (hunk 5 above)
Comment (shantanutrip) [reply test/test_utils/__init__.py:879]: Yes, this is the only occurrence - removed

**Review by arjkesh — APPROVED** [review]
